### PR TITLE
Bug 1884131: fix e2e tests to use aws cluster supplied by the CI

### PR DESCRIPTION
**Pull request by stlaz**
This PR makes the repo use the unified build machinery as most of our other repos use.

I tried to move a bit towards running the e2e-suite on our own cluster but there are problems that I don't want to be dealing with at this moment.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+# $0 - macro name
+# $1 - target name
+# $2 - image ref
+# $3 - Dockerfile path
+# $4 - context directory for image build
+# It will generate target "image-$(1)" for building the image and binding it as a prerequisite to target "images".
+$(call build-image,ocp-oauth-proxy,$(IMAGE_REGISTRY)/ocp/4.4:oauth-proxy,./Dockerfile.rhel7,.)
+
+# Run core verification and all self contained tests.
+check: | verify test-unit
```
Comment (tnozicka) [inline Makefile]: nit: a non-standard target, I don't object if you want it there

File: Makefile [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+.PHONY: check
 
 clean:
 	$(RM) ./oauth-proxy
 .PHONY: clean
+
+GO_TEST_PACKAGES :=./ ./api ./cookie ./providers
```
Comment (tnozicka) [inline Makefile]: is there any specific one you want to ignore? otherwise `./...` would seem more fitting

File: go.mod [hunk 3]
```diff
@@ -9,13 +9,14 @@ require (
...
+	github.com/openshift/library-go v0.0.0-20200116194409-e560b33797f5
 	github.com/yhat/wsutil v0.0.0-20170731153501-1d66fa95c997
-	golang.org/x/net v0.0.0-20191004110552-13f9640d40b9
+	golang.org/x/net v0.0.0-20191209160850-c0dbc17a3553
 	gopkg.in/check.v1 v1.0.0-20190902080502-41f04d3bba15 // indirect
-	k8s.io/api v0.17.0
-	k8s.io/apimachinery v0.17.0
-	k8s.io/apiserver v0.17.0
-	k8s.io/client-go v0.17.0
+	k8s.io/api v0.17.1
```
Comment (tnozicka) [inline go.mod]: these change don't belong to this commit

File: Makefile (hunk 2 above)
Comment (stlaz) [reply Makefile]: ignoring vendor/ here

File: Makefile (hunk 1 above)
Comment (stlaz) [reply Makefile]: I think I saw it somewhere else which was the sole reason I added it here, it shall be removed

File: Makefile (hunk 2 above)
Comment (tnozicka) [reply Makefile]: vendor is already ignored for a few golang releases in `./...`, did modules change it?

Comment (stlaz) [reply Makefile]: Hm, you're right, I thought I saw some vendor failures, whereas in fact these might have been failure that appear because of the `test/` directory which only contains the e2e tests and is the actual directory that needs to be avoided.

If there is a different way to have it excluded, I can do it for the repo.

For reference, the `go test ./...` failures:
```
# k8s.io/apimachinery/pkg/api/meta
vet: in k8s.io/apimachinery/pkg/api/meta, can't import facts for package "k8s.io/apimachinery/pkg/runtime": open $WORK/b157/vet.out: no such file or directory
# k8s.io/apimachinery/pkg/util/strategicpatch
vet: in k8s.io/apimachinery/pkg/util/strategicpatch, can't import facts for package "k8s.io/apimachinery/pkg/util/mergepatch": open $WORK/b389/vet.out: no such file or directory
# k8s.io/apimachinery/pkg/apis/meta/internalversion
vet: open $WORK/b381/vet.cfg: no such file or directory
open /tmp/go-build183862199/b380/vet.cfg: no such file or directory
# k8s.io/api/rbac/v1alpha1
vet: in k8s.io/api/rbac/v1alpha1, can't import facts for package "k8s.io/apimachinery/pkg/apis/meta/v1": open $WORK/b145/vet.out: no such file or directory
# k8s.io/api/rbac/v1beta1
vet: in k8s.io/api/rbac/v1beta1, can't import facts for package "k8s.io/apimachinery/pkg/apis/meta/v1": open $WORK/b145/vet.out: no such file or directory
# k8s.io/client-go/rest
vet: ../../../go/pkg/mod/k8s.io/client-go@v0.17.1/rest/client.go:27:2: could not import k8s.io/apimachinery/pkg/runtime (open $WORK/b157/_pkg_.a: no such file or directory)
open /tmp/go-build183862199/b275/vet.cfg: no such file or directory
open /tmp/go-build183862199/b373/vet.cfg: no such file or directory
open /tmp/go-build183862199/b272/vet.cfg: no such file or directory
# k8s.io/api/core/v1
can't create $WORK/b236/_pkg_.a: open $WORK/b236/_pkg_.a: no such file or directory
FAIL	github.com/openshift/oauth-proxy [build failed]
ok  	github.com/openshift/oauth-proxy/api	0.009s
ok  	github.com/openshift/oauth-proxy/cookie	0.009s
ok  	github.com/openshift/oauth-proxy/providers	0.018s
FAIL	github.com/openshift/oauth-proxy/providers/openshift [build failed]
FAIL	github.com/openshift/oauth-proxy/test/e2e [build failed]
?   	github.com/openshift/oauth-proxy/test/e2e/hello	[no test files]
ok  	github.com/openshift/oauth-proxy/util	0.091s
FAIL
```

File: go.mod (hunk 3 above)
Comment (stlaz) [reply go.mod]: made a separate commit with that change

File: Makefile (hunk 2 above)
Comment (stlaz) [reply Makefile]: added a comment and fixed it to include subdirectories

Comment (tnozicka) [reply Makefile]: I don't think that is it. `./test/e2e/` although you don't want to actually run it as you'd want to split units and e2e, **should** technically work no matter what.

I think the real cause might be related to vendoring. But I don't yet know why the tests compile separetelly but fail to build in some combinations.

Even this fails:
```
$ go clean -testcache
$ go test -mod=vendor ./util/... ./providers/...
ok  	github.com/openshift/oauth-proxy/util	0.011s
open /tmp/go-build201163669/b091/providers.test: no such file or directory
FAIL	github.com/openshift/oauth-proxy/providers [build failed]
open /tmp/go-build201163669/b128/openshift.test: no such file or directory
FAIL	github.com/openshift/oauth-proxy/providers/openshift [build failed]
FAIL
```

When I tried to clear the deps and start again, it shows incompatible version used by your code:
```
o mod init
go: creating new go.mod: module github.com/openshift/oauth-proxy
 tnozicka@lenovo-t580  ~/go/src/github.com/openshift/oauth-proxy   pr-148 ●
 go mod tidy
go: finding github.com/openshift/library-go latest
go: finding k8s.io/api v0.17.2
go: downloading k8s.io/apiserver v0.17.2
go: downloading github.com/openshift/library-go v0.0.0-20200128110723-fa93bef5ede7
go: downloading k8s.io/api v0.17.2
go: finding github.com/bitly/go-simplejson v0.5.0
go: downloading github.com/bitly/go-simplejson v0.5.0
go: extracting github.com/openshift/library-go v0.0.0-20200128110723-fa93bef5ede7
go: extracting github.com/bitly/go-simplejson v0.5.0
go: extracting k8s.io/apiserver v0.17.2
go: extracting k8s.io/api v0.17.2
go: finding k8s.io/apimachinery v0.17.2
go: downloading k8s.io/apimachinery v0.17.2
go: extracting k8s.io/apimachinery v0.17.2
go: finding github.com/18F/hmacauth latest
go: finding github.com/bmizerany/assert latest
go: finding github.com/mreiferson/go-options v1.0.0
go: downloading github.com/mreiferson/go-options v1.0.0
go: extracting github.com/mreiferson/go-options v1.0.0
go: finding golang.org/x/net latest
go: finding github.com/yhat/wsutil latest
go: downloading k8s.io/component-base v0.17.2
go: extracting k8s.io/component-base v0.17.2
github.com/openshift/oauth-proxy/providers/openshift imports
	k8s.io/apiserver/pkg/server/dynamiccertificates imports
	k8s.io/client-go/tools/events: module k8s.io/client-go@latest found (v11.0.0+incompatible), but does not contain package k8s.io/client-go/tools/events
```

Comment (tnozicka) [reply Makefile]: so let's
/hold
while we figure this out

**Comment by tnozicka** [comment]
yeah, this is gonna need some debugging, I am not sure what is broken. The fact that it passes individually could suggest linking...

```
go clean -testcache && go test -vet=off -v ./...
open /tmp/go-build833260985/b403/cookie.test: no such file or directory
open /tmp/go-build833260985/b406/providers.test: no such file or directory
open /tmp/go-build833260985/b400/api.test: no such file or directory
open /tmp/go-build833260985/b412/e2e.test: no such file or directory
open /tmp/go-build833260985/b409/openshift.test: no such file or directory
open /tmp/go-build833260985/b001/oauth-proxy.test: no such file or directory
FAIL	github.com/openshift/oauth-proxy [build failed]
FAIL	github.com/openshift/oauth-proxy/api [build failed]
FAIL	github.com/openshift/oauth-proxy/cookie [build failed]
FAIL	github.com/openshift/oauth-proxy/providers [build failed]
FAIL	github.com/openshift/oauth-proxy/providers/openshift [build failed]
FAIL	github.com/openshift/oauth-proxy/test/e2e [build failed]
?   	github.com/openshift/oauth-proxy/test/e2e/hello	[no test files]
=== RUN   TestGetCertPool
--- PASS: TestGetCertPool (0.00s)
PASS
ok  	github.com/openshift/oauth-proxy/util	0.023s
FAIL
```

File: Makefile (hunk 2 above)
Comment (stlaz) [reply Makefile]: /hold cancel
found a fun thing in the unit-tests of the `util/` repository, see commit `don't remove /tmp when unit-tests are run`

**Comment by stlaz** [comment]
/hold
free to review, I'll add the jobs to CI config once we're close to merge to verify this fix

**Comment by stlaz** [comment]
depends on https://github.com/openshift/oauth-proxy/pull/190

File: test/e2e/util.go [hunk 4]
```diff
@@ -43,26 +48,59 @@ const (
...
+func waitForSelfSAR(interval, timeout time.Duration, c kubernetes.Interface, selfSAR authorizationv1.SelfSubjectAccessReviewSpec) error {
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		res, err := c.AuthorizationV1().SelfSubjectAccessReviews().Create(
+			context.Background(),
+			&authorizationv1.SelfSubjectAccessReview{
+				Spec: selfSAR,
+			},
+			metav1.CreateOptions{},
+		)
+		if err != nil {
```
Comment (vareti) [inline test/e2e/util.go:92]: do you think `return res.Status.Allowed, err` is sufficient? I am assuming `Allowed` will be set to `false` in case of error.

File: test/e2e/proxy_test.go [hunk 5]
```diff
@@ -177,9 +177,20 @@ func TestOAuthProxyE2E(t *testing.T) {
 		// },
 	}
 
-	image := os.Getenv("TEST_IMAGE")
 	backendImage := os.Getenv("HELLO_IMAGE")
 
+	// Get the image
+	// TODO: is there a better way?
+	alertmanagerPod, err := kubeClient.CoreV1().Pods("openshift-monitoring").Get(testCtx, "alertmanager-main-0", metav1.GetOptions{})
```
Comment (vareti) [inline test/e2e/proxy_test.go:188]: maybe add a comment explaining why these values are used. I was unable to tell from the look of it.

File: test/e2e/proxy_test.go [hunk 6]
```diff
@@ -200,6 +195,7 @@ func TestOAuthProxyE2E(t *testing.T) {
 
 	t.Logf("test image: %s, test namespace: %s", image, ns)
 
+	backendImage := "nginxdemos/nginx-hello:plain-text"
```
Comment (vareti) [inline test/e2e/proxy_test.go:217]: just curious about the reason to replace the image?

Comment (stlaz) [reply test/e2e/proxy_test.go:217]: I was looking for a simple image to use and did not want to figure out how to build an image as a part of an e2e test and then upload it to the cluster's registry.

I'm open for ideas.

File: test/e2e/proxy_test.go [hunk 7]
```diff
@@ -231,60 +231,36 @@ func TestOAuthProxyE2E(t *testing.T) {
...
-			expectedErr: "did not reach upstream site",
-			pageResult:  "Hello OpenShift!\n",
-		},
+		// "upstream-ca": {
+		// 	oauthProxyArgs: []string{
+		// 		"--upstream=https://localhost:8080",
+		// 		"--upstream-ca=/etc/tls/private/upstreamca.crt",
+		// 	},
+		// 	backendEnvs: []string{"HELLO_TLS_CERT=/etc/tls/private/upstream.crt", "HELLO_TLS_KEY=/etc/tls/private/upstream.key"},
+		// 	expectedErr: "",
```
[the remark this answers is not in the document]
Comment (stlaz) [reply test/e2e/proxy_test.go:162]: sure

File: test/e2e/proxy_test.go [hunk 8]
```diff
@@ -30,163 +30,81 @@ func TestOAuthProxyE2E(t *testing.T) {
...
-				"--openshift-service-account=proxy",
 				"--upstream=http://localhost:8080",
-				"--tls-cert=/etc/tls/private/tls.crt",
-				"--tls-key=/etc/tls/private/tls.key",
-				"--tls-client-ca=/etc/tls/private/ca.crt",
-				"--cookie-secret=SECRET",
-				"--skip-provider-button",
 				`--skip-auth-regex=^/foo`,
 			},
-			backendEnvs:   []string{"HELLO_SUBPATHS=/foo,/bar"},
```
[the remark this answers is not in the document]
Comment (stlaz) [reply test/e2e/proxy_test.go:151]: this probably should've come after the test image change. Rebasing would probably bring merge conflicts, I'd rather keep the current commit order if you don't mind.

File: test/e2e/util.go (hunk 4 above)
Comment (stlaz) [reply test/e2e/util.go:92]: ... but `res` will probably be nil. The next condition is unnecessary though and can use what you're suggesting, I'll do that there.

File: test/e2e/util.go [hunk 9]
```diff
@@ -454,6 +448,91 @@ func newRequestFromForm(form *html.Node, currentURL *url.URL, user string) (*htt
...
+				break
+			}
+		}
+		_, err = oauthClient.Update(context.TODO(), oauthConfig, metav1.UpdateOptions{})
+		if err != nil {
+			t.Logf("failed to remove the test IdP from oauth/cluster: %s", err)
+		}
+
+		for i := 0; i < numUsers; i++ {
+			username := fmt.Sprintf("testuser%d", i)
```
Comment (vareti) [inline test/e2e/util.go:499]: `username` is `users[i]`?

File: test/e2e/util.go [hunk 10]
```diff
@@ -675,3 +755,36 @@ func NewClientConfigForTest(t *testing.T) *rest.Config {
...
+		clusterOperator, err := client.ClusterOperators().Get(context.TODO(), "authentication", metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			t.Logf("clusteroperators.config.openshift.io/authentication: %v", err)
+			return false, nil
+		}
+		if retry.IsHTTPClientError(err) {
+			t.Logf("clusteroperators.config.openshift.io/authentication: %v", err)
+			return false, nil
+		}
+		availableStatusIsMatch, progressingStatusIsMatch, degradedStatusIsMatch := true, true, true
```
Comment (vareti) [inline test/e2e/util.go:752]: The logic seems to be correct but do you think using this [helper](https://github.com/openshift/library-go/blob/master/pkg/config/clusteroperator/v1helpers/status.go#L121) would simplify it even further?

File: test/e2e/util.go [hunk 11]
```diff
@@ -454,6 +448,91 @@ func newRequestFromForm(form *html.Node, currentURL *url.URL, user string) (*htt
...
+
+	var users []string
+	for i := 0; i < numUsers; i++ {
+		users = append(users, fmt.Sprintf("testuser%d", i))
+	}
+
+	kubeClient.CoreV1().Secrets("openshift-config").Create(context.TODO(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nsName + "htpasswd",
```
Comment (vareti) [inline test/e2e/util.go:452]: nit: `nsName + "htpasswd"` is used multiple times. maybe use a varaible?

File: test/e2e/proxy_test.go [hunk 12]
```diff
@@ -197,15 +202,30 @@ func TestOAuthProxyE2E(t *testing.T) {
...
 	backendImage := "nginxdemos/nginx-hello:plain-text"
+	currentTestIdx := 0 // to pick the current user so that each test gets a fresh grant
 	for tcName, tc := range oauthProxyTests {
 		runOnly := os.Getenv("TEST")
 		if len(runOnly) > 0 && runOnly != tcName {
 			continue
 		}
+
 		t.Run(fmt.Sprintf("setting up e2e tests %s", tcName), func(t *testing.T) {
+			defer func() { currentTestIdx++ }()
```
[the remark this answers is not in the document]
Comment (stlaz) [reply test/e2e/proxy_test.go:217]: > t.Run internally spawns a go routine

... and waits for its result, so the more accurate example would be https://play.golang.org/p/VIrPIw8ndJt

File: test/e2e/util.go (hunk 9 above)
Comment (stlaz) [reply test/e2e/util.go:499]: I'd rather keep it like this in case `users`, reference to which we return, were tampered with:
https://play.golang.org/p/oHB69CMloDv

File: test/e2e/util.go (hunk 10 above)
Comment (stlaz) [reply test/e2e/util.go:752]: I think the current logic is the simplest for watching a subset of all the conditions while also being able to print the current operator status, but maybe you have even a simpler solution that I'm not seeing?

File: test/e2e/util.go (hunk 11 above)
Comment (stlaz) [reply test/e2e/util.go:452]: done

File: test/e2e/util.go (hunk 10 above)
Comment (vareti) [reply test/e2e/util.go:753]: maybe something similar like this? https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/test/library/cluster_operator.go#L57-L61

File: test/e2e/proxy_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (vareti) [reply test/e2e/proxy_test.go:217]: Ah.. missed that. No issues with it then. 

Although when I think about it, we can move increment operation out of `t.Run` and achieve the same without using a defer.

File: test/e2e/util.go (hunk 10 above)
Comment (stlaz) [reply test/e2e/util.go:753]: I think if we're to keep the generic state of this function, this could be harder to do.

**Comment by vareti** [comment]
@stlaz let me know once all the comments are addressed and I will tag the PR.

**Comment by stlaz** [comment]
@vareti done

**Comment by stlaz** [comment]
/retest
let's see, maybe I screwed one of the tests?

**Comment by stlaz** [comment]
/hold cancel
looks like a flake
